### PR TITLE
Rewrite SaveVulcanGSS for 27 bank GSAS

### LINKED ISSUE #21780: Enable Mantid to write a 27-bank GSAS file for VULCAN

### PR DESCRIPTION
**Pull request by wdzhou**
Description of work.

**To test:**

Fixes #21780 . 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by NickDraper** [comment]
This needs someone to test & review please

**Comment by NickDraper** [comment]
@wdzhou Wenduo, no one is picking this up to test as there are no instructions on how to test, or what the description of the work done was.

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 1]
```diff
@@ -608,11 +739,6 @@ void SaveGSS::setOtherProperties(IAlgorithm *alg,
   * @throws :: If for any reason we cannot run the algorithm
   */
 void SaveGSS::validateUserInput() const {
-  // Check whether it is PointData or Histogram
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/SaveGSS.cpp]: Why did you remove this check? Does the implementation of that have an assumption about constant number of bins?

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 2]
```diff
@@ -807,9 +940,27 @@ void SaveGSS::writeRALF_XYEdata(const int bank, const bool MultiplyByBinWidth,
...
+ * @param histo
+ * @param xye_precision
+ */
+void SaveGSS::writeSLOGdata(const size_t ws_index, const int bank,
+                            const bool MultiplyByBinWidth,
                             std::stringstream &out,
-                            const HistogramData::Histogram &histo) const {
+                            const HistogramData::Histogram &histo,
+                            const std::vector<int> &xye_precision) const {
+  // check inputs
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/SaveGSS.cpp]: Don't need to check again since it will be validated and set already.

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 3]
```diff
@@ -833,11 +987,26 @@ void SaveGSS::writeSLOGdata(const int bank, const bool MultiplyByBinWidth,
...
 
   // Write bank header
-  writeBankHeader(out, "SLOG", bank, datasize);
-  out << std::fixed << " " << std::setprecision(0) << std::setw(10) << bc1
-      << std::fixed << " " << std::setprecision(0) << std::setw(10) << bc2
-      << std::fixed << " " << std::setprecision(7) << std::setw(10) << bc3
-      << std::fixed << " 0 FXYE\n";
+  if (overwrite_std_bank_header_) {
+    // write user header only!
+    if (ws_index >= user_specified_bank_headers_.size()) {
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/SaveGSS.cpp]: Move this to `validateInputs`

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 4]
```diff
@@ -169,6 +211,9 @@ void SaveGSS::exec() {
     entry = makeStringStream();
   }
 
+  // process user special headers
+  processUserSpecifiedHeaders();
+
   // Check the user input
   validateUserInput();
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/SaveGSS.cpp]: Change this to `validateInputs()` and remove from `exec()`. Then the framework will automatically call it at the right time.

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 5]
```diff
@@ -833,11 +987,26 @@ void SaveGSS::writeSLOGdata(const int bank, const bool MultiplyByBinWidth,
   g_log.debug() << "SaveGSS(): Min TOF = " << bc1 << '\n';
 
   // Write bank header
-  writeBankHeader(out, "SLOG", bank, datasize);
-  out << std::fixed << " " << std::setprecision(0) << std::setw(10) << bc1
-      << std::fixed << " " << std::setprecision(0) << std::setw(10) << bc2
-      << std::fixed << " " << std::setprecision(7) << std::setw(10) << bc3
-      << std::fixed << " 0 FXYE\n";
+  if (overwrite_std_bank_header_) {
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/SaveGSS.cpp]: It doesn't look like the logic matches the documentation. Should it be
```
if (!user_specified_bank_headers_.empty())
  // write it out

if (!overwrite_std_bank_header_)
  // write it out
```
The current implementation cannot have both bank headers.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+        """ Declare properties
+        """
+        self.declareProperty(MatrixWorkspaceProperty("InputWorkspace", "", Direction.Input),
+                             "Focused diffraction workspace to be exported to GSAS file. ")
+
+        self.declareProperty(ITableWorkspaceProperty('BinningTable', '', Direction.Input, PropertyMode.Optional),
+                             'Table workspace containing binning parameters. If not specified, then no re-binning'
+                             'is required')
+
+        # self.declareProperty(FileProperty("BinFilename", "", FileAction.OptionalLoad, ['.dat']),
```
Comment (peterfpeterson) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py]: Remove dead code

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+        #                      "Name of a data file containing the bin boundaries in Log(TOF).")
+
+        self.declareProperty(MatrixWorkspaceProperty("OutputWorkspace", "", Direction.Output),
+                             "Name of rebinned matrix workspace. ")
+
+        self.declareProperty(FileProperty("GSSFilename","", FileAction.Save, ['.gda']),
+                             "Name of the output GSAS file. ")
+
+        self.declareProperty("IPTS", 0, "IPTS number")
```
Comment (peterfpeterson) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py]: You can allow for people leaving this blank if you use [`GetIPTS`](http://docs.mantidproject.org/nightly/algorithms/GetIPTS-v1.html).

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+        return input_workspace, binning_parameter_list, gss_file_name, output_ws_name, ipts, parm_file_name
+
+    def process_binning_param_table(self, input_workspace, bin_par_ws_name):
+        """
+        process the binning parameters given in an ITableWorkspace
+        :param input_workspace:
+        :param bin_par_ws_name:
+        :return:
+        """
+        def convert_str_to_integers(list_str):
```
Comment (peterfpeterson) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py]: Isn't this a "common" thing that already exists?

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+        process binning parameters in string.  there are two types of binning parameters that are accepted
+        1. regular x0, dx0, x1, dx1, etc or
+        2. workspace name: workspace index
+        :param bin_par_str:
+        :return:
+        """
+        if bin_par_str.count(':') == 0:
+            # parse regular binning parameters
+            terms = bin_par_str.split(',')  # in string format
+            bin_param = list()
```
Comment (peterfpeterson) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py]:
```python
try:
    bin_param = [float(term) for term in terms]
except ValueError:
￼    raise RuntimeError('Binning parameters {0} have non-float terms.'.format(bin_par_str))
```

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+            ref_tof_vec = ref_tof_ws.readX(ws_index)
+            delta_tof_vec = ref_tof_vec[1:] - ref_tof_vec[:-1]
+
+            bin_param = numpy.empty((ref_tof_vec.size + delta_tof_vec.size), dtype=ref_tof_vec.dtype)
+            bin_param[0::2] = ref_tof_vec
+            bin_param[1::2] = delta_tof_vec
+
+            self.log().warning('Binning parameters: size = {0}\n{1}'.format(len(bin_param), bin_param))
+
+        else:
```
Comment (peterfpeterson) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py]: Could be done in a `validateInputs()` method that will be called by the framework before `PyExec()`

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+            run_start_seconds = run_start.split(".")[0]
+            run_start_sub_seconds = run_start.split(".")[1]
+            self.log().warning('Run start {0} is split to {1} and {2}'.format(run_start, run_start_seconds,
+                                                                              run_start_sub_seconds))
+
+            # property run_start and duration exist
+            utctime = datetime.strptime(run_start_seconds, '%Y-%m-%dT%H:%M:%S')
+            time0 = datetime.strptime("1990-01-01T0:0:0", '%Y-%m-%dT%H:%M:%S')
+
+            # convert from string version of time to total nano second
```
Comment (peterfpeterson) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py]: The better solution is `np.datetime64`. It can be created from those time stamps and converted into nanoseconds from the start using
```python
utctime = np.datetime64(run.getProperty("run_start").value)
time0 = n.datetime64("1990-01-01T0:0:0")
total_nanosecond_start = int((utctime-time0)/np.timedelta64(1, 'ns'))
```
Then a bunch of your other code goes away.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Lots of minor changes.

File: Framework/DataHandling/inc/MantidDataHandling/SaveGSS.h [hunk 12]
```diff
@@ -161,6 +170,14 @@ class DLLExport SaveGSS : public Mantid::API::SerialAlgorithm {
   bool m_allDetectorsValid{false};
   /// Holds pointer to progress bar
   std::unique_ptr<API::Progress> m_progress{nullptr};
+  /// User specified header string
+  std::vector<std::string> user_specified_gsas_header_;
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/DataHandling/inc/MantidDataHandling/SaveGSS.h]: Done!

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 13]
```diff
@@ -96,6 +107,12 @@ void writeBankHeader(std::stringstream &out, const std::string &bintype,
 }
 } // End of anonymous namespace
 
+//----------------------------------------------------------------------------------------------
+//// Constructor
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/DataHandling/src/SaveGSS.cpp]: Done!

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 14]
```diff
@@ -186,6 +231,33 @@ void SaveGSS::exec() {
...
+//----------------------------------------------------------------------------------------------
+/** process user specified headers
+*@brief SaveGSS::processUserSpecifiedHeaders
+*/
+void SaveGSS::processUserSpecifiedHeaders() {
+
+  // user specified GSAS
+  user_specified_gsas_header_ = getProperty("UserSpecifiedGSASHeader");
+
+  if (user_specified_gsas_header_.size() == 0) {
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/DataHandling/src/SaveGSS.cpp]: Done!

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 15]
```diff
@@ -186,6 +231,33 @@ void SaveGSS::exec() {
...
+    overwrite_std_gsas_header_ = false;
+  } else {
+    overwrite_std_gsas_header_ = getProperty("OverwriteStandardHeader");
+  }
+
+  // user specified bank header
+  user_specified_bank_headers_ = getProperty("UserSpecifiedBankHeader");
+  if (user_specified_bank_headers_.size() == 0) {
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/DataHandling/src/SaveGSS.cpp]: Done!

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 16]
```diff
@@ -147,6 +164,31 @@ void SaveGSS::init() {
...
+                  "inserted before the original header");
+
+  declareProperty(
+      Kernel::make_unique<Kernel::ArrayProperty<std::string>>(
+          "UserSpecifiedBankHeader"),
+      "Each string will be used to replaced the standard GSAS bank header."
+      "Number of strings in the give array must be same as number of banks."
+      "And the order is preserved.");
+
+  declareProperty(
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/DataHandling/src/SaveGSS.cpp]: Will RALF people want to use this feature?  SNS just use SLOG though.

[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/DataHandling/src/SaveGSS.cpp]: Validator is applied.

File: Framework/DataHandling/src/SaveGSS.cpp (hunk 1 above)
Comment (wdzhou) [reply Framework/DataHandling/src/SaveGSS.cpp]: Histogram can give correct result for both histogram and point data.  Therefore, there is no need to check anymore.  The check itself is a legacy.  I tested histogram and point data from same workspace, and got same result in GSAS file.

File: Framework/DataHandling/src/SaveGSS.cpp (hunk 5 above)
Comment (wdzhou) [reply Framework/DataHandling/src/SaveGSS.cpp]: The logic is to use either user-specified header or standard header.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py (hunk 10 above)
Comment (wdzhou) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SaveVulcanGSS.py]: It is not straightforward to validate the inputs.  Either don't want to use the class member variables.

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 17]
```diff
@@ -708,38 +715,41 @@ void SaveGSS::setOtherProperties(IAlgorithm *alg,
...
-  const bool append = getProperty("Append");
-  for (const auto &filename : m_outFileNames) {
-    if (!append && doesFileExist(filename)) {
-      g_log.warning("Target GSAS file " + filename +
-                    " exists and will be overwritten.\n");
-    } else if (append && !doesFileExist(filename)) {
-      g_log.warning("Target GSAS file " + filename +
-                    " does not exist but algorithm was set to append.\n");
-    }
+  std::string output_file_name = getProperty("Filename");
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/SaveGSS.cpp]: I think this is already done for you by the property when you say it is `API::FileProperty::Save`.

File: Framework/DataHandling/src/SaveGSS.cpp [hunk 18]
```diff
@@ -708,38 +715,41 @@ void SaveGSS::setOtherProperties(IAlgorithm *alg,
...
   // Check about the user specified bank header
-  if (m_overwrite_std_bank_header &&
-      m_user_specified_bank_headers.size() !=
-          m_inputWS->getNumberHistograms()) {
-    throw std::invalid_argument("If user specifies bank header, each bank must "
-                                "have a unique user-specified header.");
+  std::vector<std::string> user_header_vec =
+      getProperty("UserSpecifiedBankHeader");
+  if (user_header_vec.size() > 0 &&
+      user_header_vec.size() != input_ws->getNumberHistograms()) {
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/SaveGSS.cpp]: Should compare to `nHist` which you already have a variable for.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Much nicer! A few minor things to change still.

**Review by peterfpeterson — APPROVED** [review]
